pkg/admission/kubequota/kubequota_admission.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	return nil
+}
+
+var _ admission.ValidationInterface = &KubeResourceQuota{}
+var _ = initializers.WantsKcpInformers(&KubeResourceQuota{})
+var _ = initializer.WantsExternalKubeInformerFactory(&KubeResourceQuota{})
+var _ = initializers.WantsKubeClusterClient(&KubeResourceQuota{})
+var _ = initializer.WantsQuotaConfiguration(&KubeResourceQuota{})
+var _ = initializers.WantsServerShutdownChannel(&KubeResourceQuota{})
+
+// Validate gets or creates a resourcequota.QuotaAdmission plugin for the logical cluster in the request and then
+// delegates validation to it.
+func (k *KubeResourceQuota) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	k.once.Do(func() {
+		m := newClusterWorkspaceDeletionMonitor(k.clusterWorkspaceInformer, k.stopQuotaAdmissionForCluster)
+		go m.Start(k.serverDone)
+	})
+
+	cluster, err := genericapirequest.ValidClusterFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	delegate, err := k.getOrCreateDelegate(cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	return delegate.Validate(ctx, a, o)
+}
+
+// getOrCreateDelegate creates a resourcequota.QuotaAdmission plugin for clusterName.
+func (k *KubeResourceQuota) getOrCreateDelegate(clusterName logicalcluster.Name) (*stoppableQuotaAdmission, error) {
+	var delegate *stoppableQuotaAdmission
+
+	k.lock.RLock()
+	delegate = k.delegates[clusterName]
+	k.lock.RUnlock()
+
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	delegate = k.delegates[clusterName]
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	// Set up a context that is cancelable and that is bounded by k.serverDone
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		// Wait for either the context or the server to be done. If it's the server, cancel the context.
+		select {
+		case <-ctx.Done():
+		case <-k.serverDone:
+			cancel()
+		}
+	}()
+
+	quotaAdmission, err := resourcequota.NewResourceQuota(k.userSuppliedConfiguration, 5, ctx.Done())
+	if err != nil {
+		return nil, err
+	}
+
+	delegate = &stoppableQuotaAdmission{
+		QuotaAdmission: quotaAdmission,
+		stopFunc:       cancel,
+	}
+
+	delegate.SetExternalKubeInformerFactory(&scopedResourceQuotaInformerFactory{
+		SharedInformerFactory: k.kubeInformers,
+		clusterName:           clusterName,
+	})
+
+	delegate.SetExternalKubeClientSet(k.kubeClusterClient.Cluster(clusterName))
+
+	delegate.SetQuotaConfiguration(k.quotaConfiguration)
+
+	if err := delegate.ValidateInitialization(); err != nil {
+		return nil, err
+	}
+
+	k.delegates[clusterName] = delegate
+
+	return delegate, nil
+}
+
+type stoppableQuotaAdmission struct {
+	*resourcequota.QuotaAdmission
+	stopFunc func()
+}
+
+func (s *stoppableQuotaAdmission) stop() {
+	s.stopFunc()
+}
+
+func (k *KubeResourceQuota) stopQuotaAdmissionForCluster(clusterName logicalcluster.Name) {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	delegate := k.delegates[clusterName]
+
```

Comment:
drop empty line

Reply:
I prefer to have a blank line before the return